virtio-devices/src/mem.rs

```diff
@@ -750,11 +752,12 @@ impl Mem {
             config.node_id = node_id;
         }
 
-        let host_fd = if let Some(f_offset) = region.file_offset() {
-            Some(f_offset.file().as_raw_fd())
-        } else {
-            None
-        };
+        let mut host_fd = None;
+        if hugepages {
+            if let Some(f_offset) = region.file_offset() {
+                host_fd = Some(f_offset.file().as_raw_fd())
```

Comment:
The code is already rather complicated. Don't add another opaque side effect driven decision. Rather if you must test for hugepages, make it explicit (which might require you to make `virtio_mem_state_change_request()` a method and store the hugepages state on the struct.

I know the kernel does not support `madvise()` with huge pages but what does that mean in terms of the host memory management. Do we even need the `madvise()` for non huge pages too

Reply:
Pushed a new version according to your comments.